csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.Nuget.cs

```diff
@@ -358,7 +374,126 @@ private void TryChangeTargetFrameworkMoniker(DirectoryInfo tempDir)
             }
         }
 
+        private static async Task ExecuteGetRequest(string address, HttpClient httpClient, CancellationToken cancellationToken)
+        {
+            using var stream = await httpClient.GetStreamAsync(address, cancellationToken);
+            var buffer = new byte[1024];
+            int bytesRead;
+            while ((bytesRead = stream.Read(buffer, 0, buffer.Length)) > 0)
+            {
+                // do nothing
+            }
+        }
+
+        private bool IsFeedReachable(string feed)
+        {
+            using HttpClient client = new();
+            var timeoutSeconds = 1;
+            var tryCount = 4;
+
+            for (var i = 0; i < tryCount; i++)
+            {
+                using var cts = new CancellationTokenSource();
+                cts.CancelAfter(timeoutSeconds * 1000);
+                try
+                {
+                    ExecuteGetRequest(feed, client, cts.Token).GetAwaiter().GetResult();
+                    return true;
+                }
+                catch (Exception exc)
+                {
+                    if (exc is TaskCanceledException tce &&
+                        tce.CancellationToken == cts.Token &&
+                        cts.Token.IsCancellationRequested)
+                    {
+                        logger.LogWarning($"Didn't receive answer from Nuget feed '{feed}' in {timeoutSeconds} seconds.");
+                        timeoutSeconds *= 2;
+                        continue;
+                    }
+
+                    // We're only interested in timeouts.
+                    logger.LogWarning($"Querying Nuget feed '{feed}' failed: {exc}");
+                    return true;
+                }
+            }
+
+            logger.LogError($"Didn't receive answer from Nuget feed '{feed}'. Tried it {tryCount} times.");
+            return false;
+        }
+
+        private bool CheckFeeds(List<FileInfo> allFiles)
+        {
+            logger.LogInfo("Checking Nuget feeds...");
+            var feeds = GetAllFeeds(allFiles);
+
+            var excludedFeeds = Environment.GetEnvironmentVariable(EnvironmentVariableNames.ExcludedNugetFeedsFromResponsivenessCheck)
```

Comment:
Is this also here to allow flexibility from a testing in production perspective?

Reply:
Yes, it's there to allow users override our checks (if we for some reason erroneously identified a feed in an earlier analysis).